lib/Sema/TypeCheckType.cpp

```diff
@@ -319,6 +320,31 @@ Type TypeChecker::resolveTypeInContext(
         }
       }
     }
+    
+    // If we found an alias type in an inherited protocol, resolve it based on our
+    // own `Self`.
+    if (alias && alias->hasUnderlyingType() && alias->getUnderlyingType()->hasArchetype()) {
+      auto typeRepr = alias->getUnderlyingTypeLoc().getTypeRepr();
+      auto identRepr = typeRepr ? dyn_cast<IdentTypeRepr>(typeRepr) : nullptr;
+
+      if (parentDC->getAsProtocolOrProtocolExtensionContext() && identRepr) {
+        auto protoSelf = parentDC->getProtocolSelf();
+        auto selfTy = protoSelf->getDeclaredType()->castTo<GenericTypeParamType>();
+        auto baseTy = resolver->resolveGenericTypeParamType(selfTy);
+        
+        if (baseTy->isTypeParameter()) {
+          for (auto name : identRepr->getComponentRange()) {
```

Comment:
Why are we using the IdentTypeRepr? The declared interface type of "alias" should be a DependentMemberType in these cases, so we can resolve it semantically.